### PR TITLE
Fix TestForStateRestore (#4476)

### PR DESCRIPTION
**Pull request by codesome**
Fixes https://github.com/prometheus/prometheus/issues/4476

### CODE REVIEW

File: rules/manager_test.go [hunk 1]
```diff
@@ -326,6 +326,21 @@ func TestForStateAddSamples(t *testing.T) {
...
+		for j > 0 {
+			if items[j-1].Labels.Hash() > items[j].Labels.Hash() {
+				items[j-1], items[j] = items[j], items[j-1]
+			}
+			j = j - 1
+		}
+	}
+}
+
 func TestForStateRestore(t *testing.T) {
```
Comment (gouthamve) [inline rules/manager_test.go]: Not part of this PR, but why the extra whitespace?

File: rules/manager_test.go [hunk 2]
```diff
@@ -326,6 +326,21 @@ func TestForStateAddSamples(t *testing.T) {
...
 }
 
+// sortAlerts sorts `[]*Alert` w.r.t. the Labels.
+// This is cleaner for small slice than adding Len(), Less(), Swap() for []*Alert.
+func sortAlerts(items []*Alert) {
+	var n = len(items)
+	for i := 1; i < n; i++ {
+		j := i
+		for j > 0 {
+			if items[j-1].Labels.Hash() > items[j].Labels.Hash() {
```
Comment (gouthamve) [inline rules/manager_test.go]: We can use `sort.Slice` here no?

Also, use https://godoc.org/github.com/prometheus/common/model#LabelSet.Before for comparing?

File: rules/manager_test.go (hunk 1 above)
Comment (codesome) [reply rules/manager_test.go]: Where exactly?

File: rules/manager_test.go (hunk 2 above)
Comment (codesome) [reply rules/manager_test.go]: Thanks for pointing me to `sort.Slice`. 

And `alert.Labels` is a slice of labels, whereas `LabelSet` is a map. Hence we will have unnecessary conversion to make?

File: rules/manager_test.go (hunk 1 above)
Comment (gouthamve) [reply rules/manager_test.go]: Right after the function start :)

File: rules/manager_test.go (hunk 2 above)
Comment (gouthamve) [reply rules/manager_test.go]: https://github.com/prometheus/prometheus/blob/134ba8e03ecd097d4f7a240c4e09a27d847ff389/pkg/labels/labels.go#L223-L225

This should be better :)

Comment (codesome) [reply rules/manager_test.go]: Ah, I missed that while looking at the API. Updated now.

File: rules/manager_test.go [hunk 3]
```diff
@@ -326,8 +326,15 @@ func TestForStateAddSamples(t *testing.T) {
 	}
 }
 
-func TestForStateRestore(t *testing.T) {
+// sortAlerts sorts `[]*Alert` w.r.t. the Labels.
+// This is cleaner for small slice than adding Len(), Less(), Swap() for []*Alert.
```
Comment (gouthamve) [inline rules/manager_test.go]: Comment needs updating.

**Comment by gouthamve** [comment]
Hah, travis failed with the other error :)

**Comment by gouthamve** [comment]
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa690d0]
goroutine 147 [running]:
github.com/prometheus/prometheus/rules.(*RecordingRule).Eval(0xc420110d90, 0xc1b940, 0xc420859f80, 0x270c56ef, 0xed30771fc, 0xee5500, 0x0, 0x0, 0xee5500, 0xc1b940, ...)
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/recording.go:73 +0xf0
github.com/prometheus/prometheus/rules.(*Group).Eval.func1(0xc1b8c0, 0xc4200a6050, 0x270c56ef, 0xed30771fc, 0xee5500, 0xc4200e2b40, 0x0, 0xc1f400, 0xc420110d90)
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:384 +0x2ec
github.com/prometheus/prometheus/rules.(*Group).Eval(0xc4200e2b40, 0xc1b8c0, 0xc4200a6050, 0x270c56ef, 0xed30771fc, 0xee5500)
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:452 +0xd0
github.com/prometheus/prometheus/rules.(*Group).run.func1()
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:230 +0x105
github.com/prometheus/prometheus/rules.(*Group).run(0xc4200e2b40, 0xc1b8c0, 0xc4200a6050)
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:243 +0x38f
github.com/prometheus/prometheus/rules.(*Manager).Update.func1.1(0xc420458eb0, 0xc4200e2b40)
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:679 +0xa7
created by github.com/prometheus/prometheus/rules.(*Manager).Update.func1
	/home/travis/gopath/src/github.com/prometheus/prometheus/rules/manager.go:674 +0x64
FAIL	github.com/prometheus/prometheus/rules	1.205s
```

Putting it here and restarted the build.

**Comment by gouthamve** [comment]
Merging this in as there is https://github.com/prometheus/prometheus/issues/4477 for the flake.
